modulus/models/unet/unet3d.py

```diff
@@ -0,0 +1,510 @@
+    -------
+        torch.Tensor: The processed output tensor.
+    """
+
+    def __init__(
+        self,
+        out_channels: int,
+        feature_map_channels: List[int],
+        model_depth: int = 3,
+        num_conv_blocks: int = 2,
+    ):
+        super().__init__()
+
+        if len(feature_map_channels) != model_depth * num_conv_blocks + 1:
+            raise ValueError(
+                "The length of feature_map_channels in the decoder block should be equal to model_depth * num_conv_blocks + 1"
+            )
+
+        self.layers = nn.ModuleList()
+        current_channels = feature_map_channels[0]
+        feature_map_channels = feature_map_channels[1:]
+
+        for depth in range(model_depth):
+            for i in range(num_conv_blocks):
+                if i == 0:
+                    self.layers.append(
+                        ConvTranspose(
+                            in_channels=current_channels,
+                            out_channels=current_channels,
+                        )
+                    )
+                    current_channels += feature_map_channels[
+                        depth * num_conv_blocks + i
+                    ]
+
+                self.layers.append(
+                    ConvBlock(
+                        in_channels=current_channels,
+                        out_channels=feature_map_channels[depth * num_conv_blocks + i],
+                    )
+                )
+                current_channels = feature_map_channels[depth * num_conv_blocks + i]
+
+        # Final convolution
+        self.layers.append(
+            ConvBlock(
+                in_channels=current_channels,
+                out_channels=out_channels,
+                activation=None,
+                normalization=None,
+            )
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        for layer in self.layers:
+            x = layer(x)
+        return x
+
+
+@dataclass
+class MetaData(ModelMetaData):
+    name: str = "UNet3D"
+    # Optimization
+    jit: bool = False
+    cuda_graphs: bool = True
+    amp: bool = True
+    # Inference
+    onnx_cpu: bool = True
+    onnx_gpu: bool = True
+    onnx_runtime: bool = True
+    # Physics informed
+    var_dim: int = 1
+    func_torch: bool = False
+    auto_grad: bool = False
+
+
+class UNet3D(Module):
+    """
+    3D U-Net model for volumetric (3D) image segmentation, featuring an encoder-decoder architecture with skip connections.
+    Default parameters are set to replicate the architecture here: https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/.
+
+    Parameters:
+    ----------
+        in_channels (int): Number of channels in the input image.
+        out_channels (int): Number of channels in the output segmentation map.
+        model_depth (int): Number of levels in the U-Net, not counting the bottleneck layer.
+        feature_map_channels (List[int]): Number of channels for each conv block in the encoder and decoder.
```

Comment:
There will be some correlation between the depth and the feature maps correct? Should we add some checks for that?

Reply:
There are already checks in the encoder and decoder for that:

https://github.com/mnabian/modulus/blob/fea-ext-3dunet/modulus/models/unet/unet3d.py#L293

https://github.com/mnabian/modulus/blob/fea-ext-3dunet/modulus/models/unet/unet3d.py#L349

Reply:
Awesome, thanks!